Add database

Sweta304 opened the pull request:
Не проходит проверка - падает на стадии сборки, локально все тесты успешно проходят

Sweta304 committed:
refactor:
- для фильмов добавлены поля жанр и рейтинг
- для ползователей добавлено поле со статусом дружбы
- в файл readme добавлена ссылка на схему БД

Sweta304 committed:
added FilmDbStorage, UserDbStorage for saving data in database

Sweta304 committed:
feature:
- в UserController добавлен метод подтверждения заявки в друзья
- создан класс Friendship в пакете model
- создан интерфейс FriendStorage, в котором доступны методы отправки и подтверждения заявки в друзья, и удаление из друзей

refactor:
- из класса User убраны методы добавления и удаления друзей
- в классе FriendStorage

Sweta304 committed:
feature:
- в UserController добавлен метод просмотра всех заявок в друзья

refactor:
- изменены типы полей в классе Friendship
- удален неиспользуемый класс FriendshipStatus

Sweta304 committed:
.

Sweta304 committed:
refactor:
- доработаны фильмы и юзеры

Sweta304 committed:
удален файл с БД

Sweta304 committed:
изменен файл api_tests.yml

Sweta304 committed:
изменен файл api_tests.yml

Sweta304 committed:
изменен файл api_tests.yml

Sweta304 committed:
changed sql files

alexache commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,52 @@
+drop table if exists FILMUSER, FILM_LIKES, USER_FRIEND, FILM, FILM_GENRE;
+
+CREATE TABLE IF NOT EXISTS mpa
```
Таблицы лучше называть заглавными буквами и без camelCase, т.е. не filmUser, а FILM_USER

alexache commented on src/main/java/ru/yandex/practicum/filmorate/DAO/FilmDbStorage.java:
```diff
@@ -0,0 +1,212 @@
+            throw new MpaNotFoundException("Рейтинга не существует. Введите значение от 1 до " + Mpa.values().length);
+        } else {
+            jdbcTemplate.update(query, film.getName(), film.getDescription(), film.getReleaseDate(), film.getDuration(), film.getRate(), film.getMpa().getId());
+            filmId = getAllFilms()
```
Довольно сложный способ получения сгенерированного айди, лучше посмотреть в сторону KeyHolder и другого метода апдейта https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/jdbc/core/JdbcTemplate.html#update-org.springframework.jdbc.core.PreparedStatementCreator-org.springframework.jdbc.support.KeyHolder-

alexache commented on src/main/java/ru/yandex/practicum/filmorate/inMemory/InMemoryFilmStorage.java:
```diff
@@ -69,4 +69,37 @@ public Film getFilmById(long id) throws FilmNotFoundException {
+
+    @Override
+    public long addLike(long filmId, long userId) throws FilmNotFoundException {
+//        Film film = getFilmById(filmId);
```
Не очень понятное изменение, зачем ломать прошлый функционал?

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/UserStorage.java:
```diff
@@ -9,7 +9,7 @@
+    User addUser(User user) throws UserAlreadyExistsException, ValidationException, UserNotFoundException;
 
     User updateUser(User user) throws UserNotFoundException, ValidationException;
 
```
Не хватает метода удаления юзера

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmStorage.java:
```diff
@@ -1,19 +1,23 @@
 
-    Film updateFilm(Film film) throws FilmNotFoundException, ValidationException;
+    Film updateFilm(Film film) throws FilmNotFoundException, ValidationException, MpaNotFoundException, GenreNotFoundException;
 
```
Не хватает метода удаления фильма

alexache commented on src/main/java/ru/yandex/practicum/filmorate/DAO/FilmDbStorage.java:
```diff
@@ -0,0 +1,212 @@
+
+    @Override
+    public Film getFilmById(long id) throws FilmNotFoundException {
+        String query = "select * from FILM where FILM_ID = ?";
```
Мне кажется, что в запросе не хватает джоина с таблицей MPA, чтобы получить значение рейтинга для фильма

alexache commented on src/main/java/ru/yandex/practicum/filmorate/DAO/FilmDbStorage.java:
```diff
@@ -0,0 +1,212 @@
+        Long filmId = resultSet.getLong("film_id");
+        List<Genres> genresList = getFilmsGenresByFilmId(filmId);
+        int mpaId = resultSet.getInt("mpa");
+        String mpaName = Mpa.values()[mpaId - 1].getValue();
```
Так нельзя делать - полагаться на значение id в таблице, которое может не совпасть с порядком в enum. Нужно делать полноценный select запрос для получения фильма и его рейтинга одним запросом

alexache commented on src/main/java/ru/yandex/practicum/filmorate/DAO/FilmDbStorage.java:
```diff
@@ -0,0 +1,212 @@
+    }
+
+    private List<Genres> getFilmsGenresByFilmId(long id) {
+        String query = "select GENRE.GENRE_ID, GENRE_NAME\n" +
```
К этому методу тоже есть вопрос,, возможно лучше ему оставить в GenresDAO, а в сервисе выше объединять данные

alexache commented on src/main/java/ru/yandex/practicum/filmorate/DAO/FilmDbStorage.java:
```diff
@@ -0,0 +1,212 @@
+            throw new FilmNotFoundException("передано некорректное кол-во фильмов");
+        }
+
+        String query = "select FILM_ID, FILM_NAME, DESCRIPTION, RELEASEDATE, DURATION, RATE, MPA from\n" +
```
Запрос выглядит слишком сложным. Чего-то вроде select * from films f  order by RATE desc limit ? должно быть достаточно

alexache commented on src/main/java/ru/yandex/practicum/filmorate/DAO/UserDbStorage.java:
```diff
@@ -0,0 +1,112 @@
+    @Override
+    public User addUser(User user) throws UserAlreadyExistsException, ValidationException, UserNotFoundException {
+        String query = "insert into FILMUSER (USER_LOGIN, USER_NAME, BIRTHDAY, EMAIL)\n" + "values (?,?,?,?)";
+        if (user.getName() == null || user.getName().isBlank() || user.getName().isEmpty()) {
```
ВОзможно валидацию лучше вынести в сервис выше

alexache commented on src/main/java/ru/yandex/practicum/filmorate/DAO/UserDbStorage.java:
```diff
@@ -0,0 +1,112 @@
+        } else {
+            jdbcTemplate.update(query, user.getLogin(), user.getName(), user.getBirthday(), user.getEmail());
+        }
+        long userId = getAllUsers().stream().filter(x -> x.getLogin().equals(user.getLogin())).findFirst().get().getId();
```
Такой же комментарий,, как и для фильмов - есть способ лучше получать сгенерированный айди

alexache commented on src/main/java/ru/yandex/practicum/filmorate/DAO/UserDbStorage.java:
```diff
@@ -0,0 +1,112 @@
+    @Override
+    public User getUserById(long id) throws UserNotFoundException {
+        String query = "select * from filmUser where user_id = ?";
+        SqlRowSet userRows = jdbcTemplate.queryForRowSet(query, id);
```
Не обязательно разделять это на 2 query - rowset и сами объекты. Можно получить все объекты и проверить их количество

Sweta304 replied:
Не смогла разобраться как это сделать

alexache commented on src/main/java/ru/yandex/practicum/filmorate/DAO/UserDbStorage.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    private Set<Long> getFriendListByUserId(long id) {
+        String query = "select FRIEND_ID\n" +
```
Возможно это тоже стоит иметь в FriendDbStorage, а не в классе юзеров

alexache commented on src/main/java/ru/yandex/practicum/filmorate/dictionary/Genre.java:
```diff
@@ -0,0 +1,10 @@
+package ru.yandex.practicum.filmorate.dictionary;
+
+public enum Genre {
```
Это не должно быть енамом, а загружаться из БД

alexache commented on src/main/java/ru/yandex/practicum/filmorate/dictionary/Mpa.java:
```diff
@@ -0,0 +1,18 @@
+
+@Getter
+public enum Mpa {
+    G("G"),
```
Это не должно быть енамом, а загружаться из БД

alexache commented on src/test/java/ru/yandex/practicum/filmorate/UserControllerTest.java:
```diff
@@ -1,155 +1,155 @@
-
-}
+//package ru.yandex.practicum.filmorate;
+//
```
Для чего закомментированный код?

alexache reviewed (changes requested):
С непроходящим билдом все же нужно разобраться, спросить совета у наставника может быть, проверить, почему может не находить процесс по порту. Может быть есть какая-та проблема у приложения. Таук же не вижу интеграционного теста, который необходим по ТЗ

Sweta304 committed:
changed database path

Sweta304 committed:
refactor:
- изменен способ получения id из таблицы
- метод получения списка друзей перенесен из UserDbStorage в FriendDbStorage
- удалены enum классы для жанров и рейтинга
- восстановлены методы из старой логики с хранением в памяти
- восстановлены тесты для хранения в памяти

feature:
- добавлены интеграционные тесты

Sweta304 committed:
- закомментировала данные для интеграционных тестов

Sweta304 committed:
fix:
- добавлены resources для папки тестов

alexache commented on src/main/java/ru/yandex/practicum/filmorate/DAO/UserDbStorage.java:
```diff
@@ -91,7 +105,7 @@ public User getUserById(long id) throws UserNotFoundException {
     private User makeUser(ResultSet resultSet, int rowNum) throws SQLException {
         Long userId = resultSet.getLong("user_id");
-        Set friends = getFriendListByUserId(userId);
+        Set friends = new FriendDbStorage(jdbcTemplate).getFriendListByUserId(userId);
```
А зачем создавать новый инстанс FriendsDbStorage? Можно же через @Autowire получить инстанс FriendsDbStorage

alexache reviewed (approved):
Спасибо за исправления
